next/test/validation/array.test.ts

```diff
@@ -0,0 +1,264 @@
+        {
+          path: ['prefixItems', 1],
+          validation: 'type',
+        },
+      ])
+    })
+  })
+
+  describe('both prefixItems and items constraints', () => {
+    const schema = {
+      type: 'array',
+      prefixItems: [{ type: 'string' }, { type: 'number' }],
+      items: { type: 'boolean' },
+    }
+
+    it('returns no errors for arrays with both prefixItems and items that match the schema', () => {
+      expect(validateSchema(['test', 42, true, false], schema)).toEqual([])
+    })
+
+    it('returns no errors for arrays that match the prefixItems constraint but have no additional items', () => {
+      expect(validateSchema(['test', 42], schema)).toEqual([])
+    })
+
+    it('returns an error for arrays that do not match the prefixItems constraint', () => {
+      expect(validateSchema(['test', 'not a number', true, false], schema)).toEqual([
+        {
+          path: ['prefixItems', 1],
+          validation: 'type',
+        },
+      ])
+    })
+
+    it('returns an error for arrays that have additional items after the prefixItems that do not match the items constraint', () => {
+      expect(validateSchema(['test', 42, true, false, 'extra'], schema)).toEqual([
+        {
+          path: ['items', 4],
+          validation: 'type',
+        },
+      ])
+    })
+  })
+
+  describe('contains', () => {
+    const schema = {
+      type: 'array',
+      contains: { type: 'number', minimum: 5 },
+    }
+
+    it('returns no errors for arrays containing at least one matching item', () => {
+      expect(validateSchema([5, 6, 7], schema)).toEqual([])
+    })
+
+    it('returns an error for arrays that do not contain any matching items', () => {
+      expect(validateSchema([1, 2, 3], schema)).toEqual([
+        {
+          path: [],
+          validation: 'contains',
+        },
+      ])
+
+      expect(validateSchema([], schema)).toEqual([
+        {
+          path: [],
+          validation: 'contains',
+        },
+      ])
+    })
+  })
+
+  it('minContains and maxContains has no effect without contains', () => {
+    const schema = {
+      type: 'array',
+      minContains: 1,
+      maxContains: 2,
+    }
+
+    expect(validateSchema([], schema)).toEqual([])
+    expect(validateSchema([1, 2, 3], schema)).toEqual([])
+  })
+
+  describe('minContains', () => {
+    const schema = {
+      type: 'array',
+      contains: { type: 'number' },
+      minContains: 2,
+    }
+
+    it('returns no errors for arrays with at least minContains items', () => {
+      expect(validateSchema([1, 2, 3], schema)).toEqual([])
+    })
+
+    it('returns an error for arrays with fewer than minContains items', () => {
+      expect(validateSchema([1], schema)).toEqual([
+        {
+          path: [],
+          validation: 'minContains',
+        },
+      ])
+    })
+  })
+
+  describe('maxContains', () => {
+    const schema = {
+      type: 'array',
+      contains: { type: 'number' },
```

Comment:
@lukad incredible work here! I can tell you we had teams asking for the `minItems` and `contains` in the past, and now seeing it implemented, is so cool! ✨